eth/dao_hard_fork.go

```diff
@@ -0,0 +1,197 @@
+		acc := statedb.GetAccount(hfData.ExtraBalances[i])
+		total.Add(total, acc.Balance())
+	}
+
+	glog.V(logger.Info).Infof("Total wei in all accounts: %v", total)
+}
+
+func prepareHardForkData(eth *Ethereum) (*DAOHardForkData, error) {
+	hfData := &DAOHardForkData{
+		TheDAOReplacementCode: theDAOReplacementCode,
+	}
+
+	theDAO, err := dao.NewDAO(theDAOAddr, NewContractBackend(eth))
+	if err != nil {
+		glog.V(logger.Info).Infof("could not instantiate DAO: %v", err)
+		return nil, err
+	}
+
+	// create list of "The DAO" and all it's children, recursively
+	daos, err := traverseDAOs(theDAO, eth)
+	if err != nil {
+		return nil, err
+	}
+	glog.V(logger.Info).Infof("DAOs traversed: %v", len(daos))
+
+	// and their extraBalance accounts
+	for i := 0; i < len(daos); i++ {
+		bal, err := daos[i].ExtraBalance(nil)
+		if err != nil {
+			glog.V(logger.Info).Infof("could not get extraBalance for DAO: %v %v", i, err)
+			return nil, err
+		}
+		hfData.ExtraBalances = append(hfData.ExtraBalances, bal)
+	}
+
+	for i := 0; i < len(daos); i++ {
+		hfData.ChildDAOs = append(hfData.ChildDAOs, daos[i].Address())
+	}
+
+	return hfData, nil
+}
+
+func traverseDAOs(d *dao.DAO, eth *Ethereum) ([]*dao.DAO, error) {
+	numberOfProposalsBig, err := d.NumberOfProposals(nil)
+	if err != nil {
+		return nil, err
+	}
+	numberOfProposals := int(numberOfProposalsBig.Uint64())
+
+	var daos []*dao.DAO
+	for i := 0; i < numberOfProposals; i++ {
+		pId := big.NewInt(int64(i))
+		proposal, err := d.Proposals(nil, pId)
+		if err != nil {
+			return nil, err
+		}
+		if proposal.NewCurator && proposal.ProposalPassed {
+			childAddr, err := d.GetNewDAOAddress(nil, pId)
+			if err != nil {
+				glog.V(logger.Info).Infof("could not get child DAO address for proposalId: %v %v", pId, err)
+				return nil, err
+			}
+			childDAO, err := dao.NewDAO(childAddr, NewContractBackend(eth))
+			if err != nil {
+				glog.V(logger.Info).Infof("could not instantiate childDAO: %x %v", childAddr, err)
+				return nil, err
+			}
+			// append this child, then all it's children (recursively)
+			daos = append(daos, childDAO)
+			children, err := traverseDAOs(childDAO, eth)
+			if err != nil {
+				glog.V(logger.Info).Infof("could not traverse children: %x %v", childAddr, err)
+				return nil, err
+			}
+			daos = append(daos, children...)
+		}
+	}
+	return daos, nil
+}
+
+// WARNING: THIS CODE DIRECTLY MODIFIES BLOCKCHAIN STATE, BYPASSING REGULAR
+//          PROTOCOL RULES. REVIEW CAREFULLY!
+// TODO: add tests!
+func applyHardFork(hfData *DAOHardForkData, eth *Ethereum) error {
+	statedb, err := eth.BlockChain().State()
+	if err != nil {
+		return err
+	}
+
+	// move all ether from all the children of the "The DAO"
+	// as well as their extraBalance accounts back to "The DAO" account
+	theDAOAcc := statedb.GetAccount(theDAOAddr)
+	bigZero := big.NewInt(0)
+
+	moveEther := func(addr common.Address) {
+		acc := statedb.GetAccount(addr)
+		bal := new(big.Int).Set(acc.Balance())
+		acc.SetBalance(bigZero)
```

Comment:
So, this will actually set the account's balance to zero, without some call to update the state, even from within a closure, because down the rabbit hole `statedb.GetAccount()->statedb.GetStateObject()` returns a pointer?..

I'm obviously new to go. :) But I feel some lines could use additional comments, even if extraneous.